common/src/main/java/com/microsoft/identity/common/internal/broker/ipc/ContentProviderStatusLoader.kt

```diff
@@ -0,0 +1,104 @@
+import com.microsoft.identity.common.java.interfaces.IPlatformComponents
+import com.microsoft.identity.common.logging.Logger
+
+/**
+ * packageManager.queryContentProviders involves an ipc call (we've seen DeadObjectException from it before)
+ *
+ * We'll avoid making unnecessary ipc request by caching them.
+ * The content provider support status should remain the same for a given broker version.
+ **/
+class ContentProviderStatusLoader(
+    private val getVersionCode: (appPackageName: String) -> String,
+    private val supportedByContentProvider: (appPackageName: String) -> Boolean,
+    private val fileManager: INameValueStorage<String>,
+) : IContentProviderStatusLoader {
+
+    companion object {
+        private val TAG = ContentProviderStatusLoader::class.java.simpleName
+        private const val SHARED_PREFERENCE_NAME = "com.microsoft.common.ipc.content.provider.query.cache"
+
+        @Throws(PackageManager.NameNotFoundException::class)
+        private fun getVersionCode(context: Context, appPackageName: String) : String {
+            val packageInfo = context.packageManager.getPackageInfo(appPackageName, 0)
+            return if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.P) {
+                packageInfo.longVersionCode.toString()
+            } else {
+                packageInfo.versionCode.toString()
+            }
+        }
+
+        @Throws(DeadObjectException::class)
+        fun supportedByContentProvider(context: Context, appPackageName: String) : Boolean {
+            val contentProviderAuthority: String =
+                ContentProviderStrategy.getContentProviderAuthority(appPackageName)
+
+            val providers: List<ProviderInfo> = context.packageManager
+                .queryContentProviders(null, 0, 0)
+
+            for (providerInfo in providers) {
+                if (providerInfo.authority != null && providerInfo.authority == contentProviderAuthority) {
+                    return true
+                }
+            }
+
+            return false
+        }
+    }
+
+    constructor(context: Context, components: IPlatformComponents): this(
+        fileManager = components.storageSupplier.getUnencryptedNameValueStore(
+            SHARED_PREFERENCE_NAME, String::class.java),
+        getVersionCode = { pkgName -> getVersionCode(context, pkgName) },
+        supportedByContentProvider = { pkgName -> supportedByContentProvider(context, pkgName) }
+    )
+
+    override fun getStatus(packageName: String) : Boolean {
+        val methodTag = "$TAG:getResult"
+        try {
+            // Construct a key for the cache.
+            val key = packageName + ":" + getVersionCode(packageName)
+
+            // Try loading from cache, return if the value is found.
+            fileManager.get(key)?.let {
+                return it.toBoolean()
+            }
+
+            val queriedResult = supportedByContentProvider(packageName)
+            fileManager.put(key, queriedResult.toString())
```

Comment:
What's the expiry/TTL on this entry?

Reply:
No expiry (could set one, but the value here shall remain unchanged since everything is defined in the manifest level)

Reply:
so next time the app upagrades, we should check it again, instead of reading from cache

Reply:
if the app upgrades, the app version should change -> the read should be a cache miss.

Reply:
@shahzaibj let me do that in a separate pr (applicable to April release onward)